inf-clojure-compliment.el

```diff
@@ -0,0 +1,112 @@
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;;
+;; Code completion using alexander-yakushev/compliment.
+
+;;; Code:
+
+(require 'edn)
+
+(defconst inf-clojure-compliment--annotations-alist
+  '((:class "c")
+    (:field "fi")
+    (:function "f")
+    (:import "i")
+    (:keyword "k")
+    (:local "l")
+    (:macro "m")
+    (:method "me")
+    (:namespace "n")
+    (:protocol "p")
+    (:protocol-function "pf")
+    (:record "r")
+    (:special-form "s")
+    (:static-field "sf")
+    (:static-method "sm")
+    (:type "t")
+    (:var "v"))
+  "Controls the abbreviations used when annotating completion candidates.
+
+Must be a list of elements with the form (TYPE . ABBREVIATION), where TYPE
+is a possible value of the candidate's type returned from the completion
+backend, and ABBREVIATION is a short form of that type.")
+
+(defun inf-clojure-compliment--get-candidate-type (symbol)
+  "Get candidate type for SYMBOL."
+  (let ((type (get-text-property 0 'type symbol)))
+    (or (cadr (assoc type inf-clojure-compliment--annotations-alist))
+        type)))
+
+(defun inf-clojure-compliment--get-candidate-ns (symbol)
+  "Get candidate ns for SYMBOL."
+  (get-text-property 0 'ns symbol))
+
+(defun inf-clojure-compliment--annotate-symbol (symbol)
+  "Return a string suitable for annotating SYMBOL.
+If SYMBOL has a text property `type` whose value is recognised, its
+abbreviation according to `cider-completion-annotations-alist' will be
+used.  If `type` is present but not recognised, its value will be used
+unaltered."
+  (let* ((type (inf-clojure-compliment--get-candidate-type symbol))
+         (ns (inf-clojure-compliment--get-candidate-ns symbol)))
+    (concat (when ns (format " (%s)" ns))
+            (when type (format " <%s>" type)))))
+
+(defun inf-clojure-compliment--parse-candidate (candidate)
+  "Get \"candidate\" from CANDIDATE.
+Put type and ns properties on the candidate"
+  (let ((entry (gethash :candidate candidate))
+        (type (gethash :type candidate))
+        (ns (gethash :ns candidate)))
+    (put-text-property 0 1 'type type entry)
+    (put-text-property 0 1 'ns ns entry)
+    entry))
+
+(defun inf-clojure-compliment--completions (response-str)
+  "Parse completions from RESPONSE-STR.
+
+Its only ability is to parse a Lisp list of candidate strings,
+every other EXPR will be discarded and nil will be returned."
+  (mapcar #'inf-clojure-compliment--parse-candidate
+          (thread-first
+              response-str
+            (inf-clojure--some)
+            (edn-read))))
+
+(setq inf-clojure-completions-fn #'inf-clojure-compliment--completions)
+(setq inf-clojure-completions-annotation-fn #'inf-clojure-compliment--annotate-symbol)
+(setq inf-clojure-completion-form "(do (require '[compliment.core :as comp]) (compliment.core/completions \"%s\"))")
```

Comment:
You don't use the `comp` alias.

Reply:
I also wonder if this should be a no-op or raise some error if the `require` fails.

Reply:
Currently everything that fails emacs'  `read` is considered an error and returns `nil`.
There is a function, I think `inf-clojure--read-or-nil`. I actually believe I should still dump errors on a file or something and not just discard them.